crates/binjs_io/src/escaped_wtf8.rs

```diff
@@ -279,3 +303,222 @@ pub fn for_print(s: &SharedString) -> SharedString {
 
     SharedString::from_string(String::from_utf8(buf).expect("Escaped string should be valid UTF-8"))
 }
+
+/// Convert the given escaped_wtf8 string to JSON.  If the given string contains
+/// any escaped lone surrogate, it's converted to unicode escape sequence.
+///
+/// This is not optimized.
+pub fn to_json(s: &SharedString) -> SharedString {
+    let input = s.as_str().as_bytes();
+
+    let mut i = 0;
+    let mut buf: Vec<u8> = Vec::with_capacity((input.len() + 2).next_power_of_two());
+    buf.push(b'\"');
+    'per_byte: while i < input.len() {
+        let unit = input[i];
+
+        if unit != LONE_SURROGATE_ESCAPE_CHAR {
+            match unit {
+                b'\\' | b'\"' => {
+                    buf.push(b'\\');
+                    buf.push(unit);
+                }
+                0x08 => {
+                    buf.push(b'\\');
+                    buf.push(b'b');
+                }
+                0x0C => {
+                    buf.push(b'\\');
+                    buf.push(b'f');
+                }
+                b'\n' => {
+                    buf.push(b'\\');
+                    buf.push(b'n');
+                }
+                b'\r' => {
+                    buf.push(b'\\');
+                    buf.push(b'r');
+                }
+                b'\t' => {
+                    buf.push(b'\\');
+                    buf.push(b't');
+                }
+                _ => {
+                    buf.push(unit);
+                }
+            }
+            i += 1;
+            continue 'per_byte;
+        }
+
+        let codepoint: u16 = (decode_hex_char(input[i + 1]) << 12)
+            | (decode_hex_char(input[i + 2]) << 8)
+            | (decode_hex_char(input[i + 3]) << 4)
+            | decode_hex_char(input[i + 4]);
+
+        if codepoint == LONE_SURROGATE_ESCAPE_CHAR as u16 {
+            buf.push(LONE_SURROGATE_ESCAPE_CHAR);
+        } else {
+            assert!(
+                codepoint >= LEAD_SURROGATE_MIN && codepoint <= TRAIL_SURROGATE_MAX,
+                "escaped codepoint should be either {:04x} or lone surrogate ({:04x}...{:04x})",
+                LONE_SURROGATE_ESCAPE_CHAR,
+                LEAD_SURROGATE_MIN,
+                TRAIL_SURROGATE_MAX
+            );
+
+            buf.extend_from_slice(b"\\u");
+            buf.push(input[i + 1]);
+            buf.push(input[i + 2]);
+            buf.push(input[i + 3]);
+            buf.push(input[i + 4]);
+        }
+
+        i += 5;
+    }
+    buf.push(b'\"');
+    SharedString::from_string(String::from_utf8(buf).expect("Escaped string should be valid UTF-8"))
+}
+
+/// Parse UTF-16 encoded JSON string and return UTF-16 string,
+/// which can contain raw lone surrogate.
+fn parse_json(input: &[u16]) -> Result<Vec<u16>, ()> {
+    assert!(input.len() >= 2);
+    assert!(input[0] == b'"' as u16);
+    assert!(input[input.len() - 1] == b'"' as u16);
+
+    let mut buf: Vec<u16> = Vec::with_capacity(input.len());
+    let mut i = 1;
```

Comment:
I think I would be more comfortable if you used a `Cursor<&[u16]>` instead of an index `i`.

Reply:
Ah, nevermind. It's self-contained, so it shouldn't really matter.